### PR TITLE
Send headers with Sec-WebSocket-* capitalization

### PR DESCRIPTION
**Pull request by qguv**
According to the RFC, the `Sec-WebSocket-*` headers should have a capital S inside the word WebSocket. This is a shame, because most other headers are canonicalized differently.

Some websocket servers are particularly picky about the capitalization. The [agar.io](https://agar.io/) server (or maybe the Cloudflare servers it's behind) will reject connections with the canonical spelling, giving a 400 in response.

You can try it. This will succeed:

    curl -v --http1.1 'https://live-arena-1cm5igj.agar.io/'
        -H 'Host: live-arena-1cm5igj.agar.io:443'
        -H 'Upgrade: websocket'
        -H 'Connection: Upgrade'
        -H 'Sec-WebSocket-Version: 13'
        -H 'Sec-WebSocket-Key: 1u4hDlp0kUjbZafgII9lLw=='

But this will fail, giving a 400 Bad Request:

    curl -v --http1.1 'https://live-arena-1cm5igj.agar.io/'
        -H 'Host: live-arena-1cm5igj.agar.io:443'
        -H 'Upgrade: websocket'
        -H 'Connection: Upgrade'
        -H 'Sec-Websocket-Version: 13'
        -H 'Sec-Websocket-Key: 1u4hDlp0kUjbZafgII9lLw=='

This patch fixes this: ws will now accept any capitalization it receives, but use the RFC-correct capitalization when sending headers either as a response or as an upgrade request.

Thanks for this library! The zero-copy functionality is really nice for us.

### CODE REVIEW

**Comment by cristaloleg** [comment]
WDYT to leave old and incorrect(😢) versions for a back compatibility and other incorrect clients?

**Comment by qguv** [comment]
@cristaloleg I don't know of any servers that require a lowercase S, but I do know of servers (like agar) that require an uppercase S. Sending the uppercase version from the RFC should allow for compatibility with the most websocket servers.

If anyone knows of a server that only accepts lowercase s headers, please comment!

**Comment by gobwas** [comment]
Hi @qguv. I think you are talking about a bug inside some particular server implementation. [HTTP RFC](https://tools.ietf.org/html/rfc2616#section-4.2) says this:
> Each header field consists of a name followed by a colon (":") and the field value. Field names are case-insensitive.

Thus it is nevermind `Sec-Websocket` or `Sec-WebSocket` or `sEc-WeBSocKET` :)

**Comment by gobwas** [comment]
But, on the other hand, the "message headers registry" [lists one](https://www.iana.org/assignments/message-headers/message-headers.xhtml) that you suggested. 🤔

**Comment by qguv** [comment]
@gobwas Yes exactly. If you look at the HTTP RFCs, it shouldn't matter. It's a shame that the WebSocket RFC breaks the convention on header normalization set in the HTTP RFC. But because I've seen servers in the wild that enforce the capital S and no servers in the wild that enforce lower-case S without accepting the capitalized version, I think it's a good idea to stay as close to the WebSocket RFC as possible.

**Comment by qguv** [comment]
Just pushed some updates. I fixed an issue that was breaking existing tests, and I added some new tests to catch issues with capitalization going forward.

**Comment by qguv** [comment]
I'd also like to point out that both Firefox and Chromium send headers with a capital S:

<summary>Firefox 66</summary><details>

```
GET / HTTP/1.1
Host: localhost:8888
User-Agent: Mozilla/5.0 (X11; Fedora; Linux x86_64; rv:66.0) Gecko/20100101 Firefox/66.0
Accept: */*
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate
Sec-WebSocket-Version: 13
Origin: null
Sec-WebSocket-Extensions: permessage-deflate
Sec-WebSocket-Key: nMTnCo98JAkQQkkmozr3GA==
DNT: 1
Connection: keep-alive, Upgrade
Pragma: no-cache
Cache-Control: no-cache
Upgrade: websocket
```
</details>

<summary>Chromium 73</summary><details>

```
GET / HTTP/1.1
Host: localhost:8888
Connection: Upgrade
Pragma: no-cache
Cache-Control: no-cache
User-Agent: Mozilla/5.0 (X11; Fedora; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/73.0.3683.86 Safari/537.36
Upgrade: websocket
Origin: chrome-search://local-ntp
Sec-WebSocket-Version: 13
Accept-Encoding: gzip, deflate, br
Accept-Language: en-US,en;q=0.9
Sec-WebSocket-Key: 2fym2MfB48lWtBYavpsdOg==
Sec-WebSocket-Extensions: permessage-deflate; client_max_window_bits
```
</details>

You can test this by running a simple ncat server:

    nc -lp 8888

and creating a new websocket in an about:blank tab in your browser of choice:

    let ws = new WebSocket('ws://localhost:8888');

You should see the headers in `nc` stdout.

File: http.go [hunk 1]
```diff
@@ -39,14 +39,21 @@ var (
...
 var (
-	headerHost          = textproto.CanonicalMIMEHeaderKey("Host")
-	headerUpgrade       = textproto.CanonicalMIMEHeaderKey("Upgrade")
-	headerConnection    = textproto.CanonicalMIMEHeaderKey("Connection")
-	headerSecVersion    = textproto.CanonicalMIMEHeaderKey("Sec-Websocket-Version")
-	headerSecProtocol   = textproto.CanonicalMIMEHeaderKey("Sec-Websocket-Protocol")
-	headerSecExtensions = textproto.CanonicalMIMEHeaderKey("Sec-Websocket-Extensions")
-	headerSecKey        = textproto.CanonicalMIMEHeaderKey("Sec-Websocket-Key")
-	headerSecAccept     = textproto.CanonicalMIMEHeaderKey("Sec-Websocket-Accept")
+	headerHost       = "Host"
```
Comment (gobwas) [inline http.go]: Could you please add the same `Canonical` version for these headers for better symmetry?

File: server_test.go [hunk 2]
```diff
@@ -332,10 +366,13 @@ func TestHTTPUpgrader(t *testing.T) {
...
 				}
-				test.req.Header.Set(headerSecKey, string(nonce))
+				if test.secKeyHeader == "" {
+					test.secKeyHeader = headerSecKey
+				}
+				test.req.Header[test.secKeyHeader] = []string{string(nonce)}
 			}
 			if test.err == nil {
-				test.res.Header.Set(headerSecAccept, string(makeAccept(test.nonce)))
+				test.res.Header[headerSecAccept] = []string{string(makeAccept(test.nonce))}
```
Comment (gobwas) [inline server_test.go]: Why these changes are needed?

**Comment by gobwas** [comment]
Hi @narqo, commented few lines in the diff, other things looks good.

File: http.go (hunk 1 above)
Comment (qguv) [reply http.go]: Done!

File: server_test.go (hunk 2 above)
Comment (qguv) [reply server_test.go]: The `(http.Header) Set()` method calls `(textproto.MIMEHeader) Set()` which [canonicalizes the header key](https://golang.org/src/net/textproto/header.go?s=705:732) before setting the header. This causes tests to fail because the expected response (forced to lowercase-s headers) does not match the actual response (which now contains uppercase-S headers). We can bypass canonicalization by setting the key directly in the map that the textproto.MIMEHeader is based on.

**Comment by gobwas** [comment]
Hi @qguv, sorry for the late reply!
Merged this PR finally. Thank you a lot 👍

**Comment by gobwas** [comment]
Also, I am about bump the *patch* version to `1.0.1`. Does it looks fine for you guys @qguv @cristaloleg?

**Comment by qguv** [comment]
Thanks @gobwas! Yes, the merge looks good.

**Comment by qguv** [comment]
Hey @gobwas, are you still planning to tag a new release?

**Comment by gobwas** [comment]
Hi there! Yes, I am. But currently Im on vacation, sorry. Will do that at Sunday probably.

**Comment by gobwas** [comment]
@qguv done! 🎆
